src/Features/CSharp/Portable/ConvertLinq/CSharpConvertLinqQueryToForEachProvider.cs

```diff
@@ -0,0 +1,920 @@
+                                    joinClause.InExpression,
+                                    capitalize: false,
+                                    _cancellationToken);
+                                SyntaxToken variable = GetFreeSymbolNameAndMarkUsed(expressionName);
+                                extraStatementToAddAbove = CreateLocalDeclarationStatement(variable, joinClause.InExpression);
+
+                                // Replace YY() with yy declared above.
+                                expression = SyntaxFactory.IdentifierName(variable);
+                            }
+
+                            // Output for the join
+                            // var yy == YY(); this goes to extraStatementToAddAbove
+                            // ...
+                            // foreach (var y in yy)
+                            // {
+                            //  if (object.Equals(x, y))
+                            //  {
+                            //      ...
+                            return SyntaxFactory.Block(
+                                SyntaxFactory.ForEachStatement(
+                                    joinClause.Type ?? VarNameIdentifier,
+                                    joinClause.Identifier,
+                                    expression,
+                                    SyntaxFactory.Block(
+                                        SyntaxFactory.IfStatement(
+                                            SyntaxFactory.InvocationExpression(
+                                                SyntaxFactory.MemberAccessExpression(
+                                                    SyntaxKind.SimpleMemberAccessExpression,
+                                                    SyntaxFactory.PredefinedType(SyntaxFactory.Token(SyntaxKind.ObjectKeyword)),
+                                                    SyntaxFactory.IdentifierName(nameof(object.Equals))),
+                                                SyntaxFactory.ArgumentList(SyntaxFactory.SeparatedList(
+                                                    new[] {
+                                                        SyntaxFactory.Argument(joinClause.LeftExpression),
+                                                        SyntaxFactory.Argument(joinClause.RightExpression.WithoutTrailingTrivia())}))),
+                                            statement)))).WithAdditionalAnnotations(Simplifier.Annotation);
+                        }
+                    case SyntaxKind.SelectClause:
+                        // This is not the latest Select in the Query Expression
+                        // There is a QueryBody with the Continuation as a parent.
+                        var selectClause = (SelectClauseSyntax)node;
+                        var identifier = ((QueryBodySyntax)selectClause.Parent).Continuation.Identifier;
+                        return AddToBlockTop(CreateLocalDeclarationStatement(identifier, selectClause.Expression), statement);
```

Comment:
>CreateLocalDeclarationStatement [](start = 45, length = 31)

We might want to use converted type of selectClause.Expression for this local if it is different from Type and doesn't include Anonymous Types

Reply:
This doesn't have to be addressed in this PR

---
In reply to: [178411441](https://github.com/dotnet/roslyn/pull/25362#discussion_r178411441) [](ancestors = 178411441)

Reply:
1. added checks for internal anonymous types
2. added declaration from Converted types in intermediate selects

---
In reply to: [178411739](https://github.com/dotnet/roslyn/pull/25362#discussion_r178411739) [](ancestors = 178411739,178411441)